src/Interpreters/Aggregator.cpp

```diff
@@ -1060,44 +1062,35 @@ void NO_INLINE Aggregator::executeImpl(
     ColumnRawPtrs & key_columns,
     AggregateFunctionInstruction * aggregate_instructions,
     bool no_more_keys,
+    bool all_keys_are_const,
     AggregateDataPtr overflow_row) const
 {
     typename Method::State state(key_columns, key_sizes, aggregation_state_cache);
 
+    auto call_execute_impl_batch = [&]<bool b1, bool b2, bool b3, bool b4>()
+    {
+        executeImplBatch<b1, b2, b3, b4>(method, state, aggregates_pool, row_begin, row_end, aggregate_instructions, overflow_row);
+    };
+
+    bool use_compiled_functions = false;
+#if USE_EMBEDDED_COMPILER
+    use_compiled_functions = compiled_aggregate_functions_holder && !hasSparseArguments(aggregate_instructions);
+#endif
+
     if (!no_more_keys)
     {
         /// Prefetching doesn't make sense for small hash tables, because they fit in caches entirely.
         const bool prefetch = Method::State::has_cheap_key_calculation && params.enable_prefetch
             && (method.data.getBufferSizeInBytes() > min_bytes_for_prefetch);
-
-#if USE_EMBEDDED_COMPILER
-        if (compiled_aggregate_functions_holder && !hasSparseArguments(aggregate_instructions))
-        {
-            if (prefetch)
-                executeImplBatch<false, true, true>(
-                    method, state, aggregates_pool, row_begin, row_end, aggregate_instructions, overflow_row);
-            else
-                executeImplBatch<false, true, false>(
-                    method, state, aggregates_pool, row_begin, row_end, aggregate_instructions, overflow_row);
-        }
-        else
-#endif
-        {
-            if (prefetch)
-                executeImplBatch<false, false, true>(
-                    method, state, aggregates_pool, row_begin, row_end, aggregate_instructions, overflow_row);
-            else
-                executeImplBatch<false, false, false>(
-                    method, state, aggregates_pool, row_begin, row_end, aggregate_instructions, overflow_row);
-        }
+        BoolArgsToTemplateArgsDispatcher<decltype(call_execute_impl_batch)>::call(call_execute_impl_batch, no_more_keys, use_compiled_functions, prefetch, all_keys_are_const);
```

Comment:
Why is it better? I didn't understand why we need this class.

Reply:
Otherwise we will need to write ~16 if/else branches to check all possibilities (maybe some of them won't actually appear, but still a lot).
Like https://pastila.nl/?01f3324d/a170ed3f747f063f81f5821188dd8aa3#B7nUAYOQzbzMvGUQ6wClTw==
And imho do it in a few lines with some class dispatcher will be better. But you can disagree and I can bring back this batch or if/else

Reply:
I do not like dispatchers as well. I think the issue is with executeImplBatch and we should refactor it somehow.